app/models/concerns/raceable.rb

```diff
@@ -0,0 +1,62 @@
+module Raceable
+  extend ActiveSupport::Concern
+
+  included do
+    # These can only be accessed from classes that include this module, not from the module itself
+    enum visibility: {public: 0, invite_only: 1, secret: 2}, _suffix: true
+
+    OPEN_ENTRY = 'Open Entry'.freeze
+    IN_PROGRESS = 'In Progress'.freeze
+    ENDED = 'Ended'.freeze
+
+    belongs_to :owner, foreign_key: :user_id, class_name: 'User'
+    has_many :entrants, as: :raceable, dependent: :destroy
+    has_many :users, through: :entrants
+    has_one :chat_room, dependent: :destroy
+    has_many :chat_messages, through: :chat_room
+
+    has_secure_token :auth_token
+
+    validates :status_text, presence: true
+
+    def started?
+      started_at.present?
+    end
+
+    def in_progress?
+      started? && !finished?
+    end
+
+    def finished?
+      entrants.all? { |entrant| entrant.finished? || entrant.forfeited? }
```

Comment:
Is there a way to accomplish this in a query, without selecting all entrants into memory first?

Reply:
We can probably query based on the status text, but that implies that we only ever have a few fixed status texts.  Another solution would be to have a finished at or closed at field directly on the race.  If you wanted the closed chat behavior i talked about in the other thing this could also be based on that finished at timestamp and could remove the chat room object if you still wanted the locking behavior.

Reply:
Not sure I follow your first sentence -- I mean can we do something like
```ruby
!Entrant.where(race: race, finished_at: nil).or(Entrant.where(race: race, forfeited_at: nil)).any?
```

Reply:
Ah, thought you meant without querying entrants for some reason, not sure why though.  Does this looks alright?

```
    def finished?
      started? && Entrant.where(raceable: self, finished_at: nil, forfeited_at: nil).none?
    end
```

I tried to do with by `join`ing on raceable so that it could check `started_at` in the db as well, but that was throwing a polymorphic preload error.

Reply:
Looks good to me! 1-2 queries is better than n. Also we may be able to clean the syntax a tiny bit by replacing `Entrant.where(raceable: self...` with `entrants.where(...` (I didn't catch that when I was suggesting it)